samples/interleave/sample.go

```diff
@@ -0,0 +1,803 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"flag"
+	"fmt"
+	"math/rand"
+	"os"
+	"strings"
+	"time"
+
+	spannergorm "github.com/googleapis/go-gorm"
+	_ "github.com/googleapis/go-sql-spanner"
+	"gorm.io/datatypes"
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+)
+
+type Singer struct {
+	gorm.Model
+	FirstName sql.NullString
+	LastName  string
+	// FullName is generated by the database. The '->' marks this a read-only field.
+	FullName string `gorm:"->;type:STRING(100) AS (ARRAY_TO_STRING([first_name, last_name], \" \")) STORED;"`
+	Active   bool
+	Albums   []Album
+}
+
+type Album struct {
+	gorm.Model
+	Title           string
+	MarketingBudget sql.NullFloat64
+	ReleaseDate     sql.NullTime
+	CoverPicture    []byte
+	SingerId        int64
+	Singer          Singer
+	Tracks          []Track `gorm:"foreignKey:id"`
+}
+
+// Track is interleaved in Album. The ID column is both the first part of the primary key of Track, and a
+// reference to the Album that owns the Track.
+type Track struct {
+	gorm.Model
+	TrackNumber int64 `gorm:"primaryKey;autoIncrement:false"`
+	Title       string
+	SampleRate  float64
+	Album       Album `gorm:"foreignKey:id"`
+}
+
+type Venue struct {
+	gorm.Model
+	Name        string
+	Description string
+}
+
+type Concert struct {
+	gorm.Model
+	Name      string
+	Venue     Venue
+	VenueId   int64
+	Singer    Singer
+	SingerId  int64
+	StartTime time.Time
+	EndTime   time.Time
+}
+
+var rnd = rand.New(rand.NewSource(time.Now().UnixNano()))
+
+// Run with `go run sample.go` to run the sample application.
+func main() {
+	// TODO(developer): Uncomment if your environment does not already have default credentials set.
+	// os.Setenv("GOOGLE_APPLICATION_CREDENTIALS", "/path/to/credentials.json")
+
+	// TODO(developer): Replace defaults with your project, instance and database if you want to run the sample
+	//                  without having to specify any command line arguments.
+	project := flag.String("project", "my-project", "The Google Cloud project of the Cloud Spanner instance")
+	instance := flag.String("instance", "my-instance", "The Cloud Spanner instance to connect to")
+	database := flag.String("database", "my-database", "The Cloud Spanner database to connect to")
+
+	flag.Parse()
+	if err := runSample(*project, *instance, *database); err != nil {
+		fmt.Printf("Failed to run sample: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+func runSample(project, instance, database string) error {
+	return RunSample("projects/" + project + "/instances/" + instance + "/databases/" + database)
+}
+
+func RunSample(connString string) error {
```

Comment:
As discussed offline: We should have an integration test that runs this sample to verify that it works and continues to work.

Reply:
Added tests